### DISCUSSION TITLE
Adding externally created OASIS file as object within the chip level

### DISCUSSION THREAD

**Discussion by zachparrott** [iqm-finland/KQCircuits#83] [Q&A] [answered]
Currently I am working with an intermediate workflow where I have some `base metal` layer geometries generated through an older workflow, but want to use alongside KQCircuits junctions and the rest of the mask generation workflow. I have discovered from the QCD group OASIS files in the junction library folder I can similarly place my desired file (let's call it `QQQ.oas`) in the qubits folder and it will be accessible in the GUI KLayout side of things. 

What I would like to really do is then write a `Chip` class file where I combine this `QQQ.oas` file with various placed junctions and say launchpads. But, it is not clear what the best way to have the OASIS file accessible in my `Chip` class file. 

From `export_cells_as_code.py` script I see I can have something like:
> ```
> view = KLayoutView()
> layout = view.layout
> cell = layout.create_cell("QQQ", "Qubits Library")
> view.insert_cell(cell, pya.DTrans(0,0))
> ```
But that code block doesn't work well if I define a new Chip class. 

From the example file `macros\masks\m001.py` I see that the `add_chip` method from `MaskSet` works fine with Chip class objects or OASIS files just fine. And separately there is the file `util\replaces_squids.py` which can take in either a SQUID calss object or .gds/.oas files. 

Does anyone have recommendation for how to do this within a Chip class I am writing using an OASIS file that functions as an externally drawn Qubit object?

**Comment by qpavsmi** [comment, marked as answer]
Hi! Sorry for the delay in response.

There are many way to achieve what you want, but I think this approach fits your workflow the best.

Assuming you have qubit geometry saved in external .oas file. It helps if the file name and the name of the "Top cell" when file is loaded in KLayout is one and the same. I have a qubit named "somequbit.oas" and in this screenshot I opened the file in KLayout, you should see that the top cell has same string as the file name
![image](https://github.com/iqm-finland/KQCircuits/assets/109152229/47ff84bb-1e02-4bdd-acc1-71c996e77f67)

Now place the file in the `klayout_package/python/kqcircuits/qubits` folder, among other qubits.

Then in your chip class code you are able to place the qubit using this kind of code:
```
from kqcircuits.chips.chip import Chip
from kqcircuits.qubits.qubit import Qubit
from kqcircuits.junctions.junction import Junction
from kqcircuits.pya_resolver import pya
from kqcircuits.util.parameters import add_parameters_from

@add_parameters_from(Junction, "junction_type")
class MyChip(Chip):
    """"""
    def build(self):
        """Produces a SingleXmons PCell."""
        self.produce_junction_tests(self.junction_type) # Just some standard chip content, remove if not needed
        self.launchers = self.produce_launchers("SMA8") # Just some standard chip content, remove if not needed

        # Relevant part - load "somequbit" from the Qubit library
        qubit_cell = self.layout.create_cell("somequbit", Qubit.LIBRARY_NAME)
        self.insert_cell(qubit_cell, pya.DTrans(0, False, 5000, 5000), "qb_0")  # Places qubit at the center of the chip
```

If you're specifically working with qubits, then there is a separate consideration on how you are planning to include junctions for this qubit. With code generated qubits we use `produce_squid` method so that the junction type can be easily swapped out at the same location, and junction's parameters could be modified easily. For statically defined geometry the junction geometry probably needs to be static as well.

Let me know if the solution above works for you or if there are other considerations I can help with.

**Reply by zachparrott** [reply]
Thank you Pavel that works as intended for now!

**Reply by zachparrott** [reply]
If anyone comes across this discussion, this method works as is for mask export using the KLayout macros. If you want to run your mask export as a console script you will need to add the following:

> from kqcircuits.util.library_helper import load_libraries
> load_libraries(path=Qubit.LIBRARY_PATH)
